mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -920,6 +924,14 @@ object ALS extends DefaultParamsReadable[ALS] with Logging {
     require(intermediateRDDStorageLevel != StorageLevel.NONE,
       "ALS is not designed to run without persisting intermediate RDDs.")
 
+    // Indeterminate rating RDD causes inconsistent in/out blocks in case of rerun.
+    // It can cause runtime error when matching in/out user/item blocks.
+    if (ratings.outputDeterministicLevel == DeterministicLevel.INDETERMINATE) {
```

Comment:
Wow, I didn't even know this existed.
I get it, but, this is true of so many operations in Spark. Why is ALS special?

Reply:
It is because how ALS is implemented.

ALS uses the training data RDD, to make user/item in/out blocks, like:

training RDD -> user in/out block
training RDD -> item in/out block

Later, it matches user in block with item out block, and item in block with user out block.

If the training RDD is indeterminate, any rerun of its tasks can produce different output in such user/item blocks. In any iteration, if such rerun happens, mismatch hits and user/item index can't find correspond slot in user/item factors.

Reply:
if I understanding this correctly, mismatch -> failure is only one possible outcome? it could also end up matching the wrong user/item before the index is wrong? that seems more subtle and much harder to detect

Reply:
Yes, it is possible. It is just hard to tell if it is the case during ALS fitting time.